### DISCUSSION TITLE
Vim `cc` equivalent (replace line content)

### DISCUSSION THREAD

**Discussion by xJonathanLEI** [helix-editor/helix#4507] [Q&A]
I tried searching but couldn't find anything relevant. Basically what I'm looking for is the same thing as `cc` and `S` in Vim:

> ["x]cc: Delete [count] lines [into register x] and start insert linewise.  If 'autoindent' is on, preserve the indent of the first line.

This is useful when I want to **rewrite** the whole line. The closest alternative in Helix that I'm aware of is `X_c`, which selects the line (including EOL), shrinks the selection to trim whitespaces, and replace. However this has drawbacks:

- it's much harder to press compared to `cc` or even `S` (solvable thru remapping, but not user friendly)
- it doesn't work on empty lines - the shrinking part doesn't work and the whole line is deleted instead.

**Comment by the-mikedavis** [comment]
There's some prior discussion in the discussions tab:

https://github.com/helix-editor/helix/discussions/2783
https://github.com/helix-editor/helix/discussions/4225

You can emulate it by creating a binding for `["goto_line_start", "select_mode", "goto_line_end", "change_selection"]`

**Reply by xJonathanLEI** [reply]
But this wouldn't work for empty lines right? This would just delete the line before turning into insert mode.

**Reply by xJonathanLEI** [reply]
Just tested and can confirm it doesn't work on empty lines. The line gets deletes because helix always selects the EOL character when the line is empty. So we still don't have an exact equivalent of `cc`: something that I can use to rewrite the line without having to think about whether the line is empty or not.

**Comment by dimas-cyriaco** [comment]
I think this would work:

```
c = { c = ["extend_to_line_bounds", "trim_selections", "change_selection"] }
```

**Reply by xJonathanLEI** [reply]
But this wouldn't work for empty lines right? This would just delete the line before turning into insert mode.

**Reply by ksandvik** [reply]
Tested c = { c = ["extend_to_line_bounds", "trim_selections", "change_selection"] } - works on empty lines, at least for me.

**Reply by xJonathanLEI** [reply]
Hmm for by "empty" I meant _really_ empty lines. The ones with whitespaces don't count.

I think this is really the same as `X_c`: it deletes the line and put cursor on the start of next line instead.

**Reply by ksandvik** [reply]
Yes, my mistake, the intent was to start typing on an empty line, then again in those cases isn't 'i' the intended command?

**Reply by xJonathanLEI** [reply]
Yes, `i` would work, but:

- Now I have to think about whether I'm rewriting an empty line or not, which is a little bit of extra mental overhead - not as good as `cc` which always just works.

- If you do `cc` in vim, even on an empty line, it clears the line and puts your cursor on the appropriate indentation. Pressing `i` on an empty line does not.

**Reply by dimas-cyriaco** [reply]
> But this wouldn't work for empty lines right? This would just delete the line before turning into insert mode.

Sorry, I didn't test this scenario. It does not work on empty lines. It deletes the whole line.

**Comment by xJonathanLEI** [comment]
If we confirm there's currently no way to actually do the same as `cc` or `S`, can we add a command for it? Should be very easy to implement if we ignore the auto-indent thing for now. Would like to know the maintainers' thoughts on this. Thanks! @the-mikedavis

In any case, I will implement this on my own fork and start using it. Maybe I should just open the PR regardless? Not sure if unsolicited PRs are considered too annoying here though.

**Reply by the-mikedavis** [reply]
I don't think there's a way to implement this that respects the select-then-act paradigm because of the behavior on an empty line. Breaking select-then-act for vim compatibility is very distasteful IMO.

**Reply by xJonathanLEI** [reply]
It's not for "vim compatability" though. The vim feature solves a real problem, and it's for solving that same problem.

But I guess that's a no then. I think it's a bit unfortunate though.

**Reply by rumpelsepp** [reply]
I am experimenting with `X_dO`, seems to work. Currently I am trying to assign this to `L`. Wish me luck. :D

edit:

I came up with:

```
[keys.normal]
L = ["extend_to_line_bounds", "delete_selection_noyank", "open_above"]
```

**Reply by xJonathanLEI** [reply]
That's the same thing I eventually came up with (but didn't post here) lol. The good news is the team seems to be going to change `c` to _not_ remove EOL if the whole line is selected. What constitues a "whole line" is unclear but this would probably make "replace line" usable without remapping.

**Reply by rumpelsepp** [reply]
But it wouldn't handle proper indenting, would it? The `open_above` thing indents the cursor if needed, that's what `cc` in vim does as well.
